### DISCUSSION TITLE
BigQueryInsertJobOperator _handle_job_error() as static method?

### DISCUSSION THREAD

**Discussion by ernvalentino** [apache/airflow#33109] [General] [answered]
I am wondering the purpose of having a method to handle BigQuery insert [job errors](https://github.com/apache/airflow/blob/2.5.3/airflow/providers/google/cloud/operators/bigquery.py#L2613-L2616) being static.
For what I intended, the purpose of having that method could be to extend it for your own need and to be flexible enough to handle errors from BigQuery as you want, because otherwise I don't see the need of having a method that just throw an exception.

But if I got the intention correctly, then why the method is static? To provide the users the flexibility to handle errors, maybe is not enough to only access the job object in that method, but rather one would access class variables or even more extending the operator introducing other class variables that could be used specifically for handling errors.

I got this idea to extend the operator to handle BigQuery load errors when doing batch loads and when setting `max_bad_records` greater than 0, to inspect the errors returned by the job and maybe send them elsewhere (a dead letter queue). But since this method is static I can't do it just extending the  `_handle_job_error()` method - even though still possible to do it in other ways -, but I would like to understand better the goal of that static method then because I am not really getting the usefulness as it is now.

Thanks for any feedback!

**Comment by hussein-awala** [comment, marked as answer]
> But if I got the intention correctly, then why the method is static? 

Simply if the method is not bound to the class instance, we could define it as class or static method. The difference between these two types is that the the class method is bound to the class, and has an access to its state where it takes a class parameter that points to the class. But since we don't need this access in our case (we process a job provided as parameter), the best type for this method is static. So the answer to your question is that we define it as a static method because it's a utility function. And as you can see, and it's defined as a method to avoid code duplication, where we call it from different place.

**Reply by ernvalentino** [reply]
I see that the method only needs the job object and thus doesn't need the operator object state, but I was wondering if there could be a vision in which to handle failures the method could access also the operator object or this is totally out of scope of the method. It seems that is out of scope based on your reply because is just an auxiliary function, while I got its purpose differently at first sight. Thanks for the explanation!

**Reply by ernvalentino** [reply]
My initial intuition was wrong, the method isn't there to be extended as per user needs but rather for better code maintainability.
